gtdynamics/optimizer/ConstraintManifold.cpp

```diff
@@ -0,0 +1,275 @@
+void ConstraintManifold::compute_basis() {
+  if (params_->basis_type == Params::BasisType::KERNEL) {
+    compute_basis_kernel();
+  } else if (params_->basis_type == Params::BasisType::SPECIFY_VARIABLES) {
+    compute_basis_specify_variables();
+  }
+}
+
+/* ************************************************************************* */
+const gtsam::Value& ConstraintManifold::recover(const gtsam::Key key,
+                                                ChartJacobian H) const {
+  if (H) {
+    // choose the corresponding rows in basis
+    *H = basis_.middleRows(var_location_.at(key), var_dim_.at(key));
+  }
+  return values_.at(key);
+}
+
+/* ************************************************************************* */
+ConstraintManifold ConstraintManifold::retract(const gtsam::Vector& xi,
+                                               ChartJacobian H1,
+                                               ChartJacobian H2) const {
+  // Compute delta for each variable and perform update.
+  gtsam::Vector x_xi = basis_ * xi;
+  gtsam::VectorValues delta;
+  for (const Key& key : cc_->keys) {
+    delta.insert(key, x_xi.middleRows(var_location_.at(key), var_dim_.at(key)));
+  }
+  gtsam::Values new_values = values_.retract(delta);
+
+  // Temporarily set jacobian as 0 since they are not used for optimization.
+  if (H1) H1->setZero();
+  if (H2) H2->setZero();
+
+  // Satisfy the constraints in the connected component.
+  return createWithNewValues(new_values, true);
+}
+
+/* ************************************************************************* */
+gtsam::Vector ConstraintManifold::localCoordinates(const ConstraintManifold& g,
+                                                   ChartJacobian H1,
+                                                   ChartJacobian H2) const {
+  Eigen::MatrixXd basis_pinv =
+      basis_.completeOrthogonalDecomposition().pseudoInverse();
+  gtsam::VectorValues delta = values_.localCoordinates(g.values_);
+  gtsam::Vector xi_base = Vector::Zero(base_dim_);
+  for (const auto& it : delta) {
+    const Key& key = it.first;
+    xi_base.middleRows(var_location_.at(key), var_dim_.at(key)) = it.second;
+  }
+  gtsam::Vector xi = basis_pinv * xi_base;
+
+  // Temporarily set jacobian as 0 since they are not used for optimization.
+  if (H1) H1->setZero();
+  if (H2) H2->setZero();
+
+  return xi;
+}
+
+/* ************************************************************************* */
+void ConstraintManifold::print(const std::string& s) const {
+  std::cout << (s.empty() ? s : s + " ") << "ConstraintManifold" << std::endl;
+  values_.print();
+}
+
+/* ************************************************************************* */
+bool ConstraintManifold::equals(const ConstraintManifold& other,
+                                double tol) const {
+  return values_.equals(other.values_, tol);
+}
+
+
+/* ************************************************************************* */
+gtsam::Values ConstraintManifold::retract_uopt(
```

Comment:
All class methods should be in camel case. Can you please follow the Google C++ style guide?